insights/combiners/os_release.py

```diff
@@ -247,54 +249,44 @@ class OSRelease(object):
-        True
-        >>> osr.reasons['kernel']
-        '4.18.0-372.19.1.el8_6uek.x86_64'
-        >>> osr.reasons['kernel_vendor'] == 'Oracle America'
+        >>> 'kernel' in osr.reasons.keys()
+        False
+        >>> 'faulty_packages' in osr.reasons.keys()
         True
         >>> 'glibc-2.28-211.el8' in osr.reasons['faulty_packages']
         True
     """
     def __init__(self, uname, dmesg, rpms, osr, rhr):
-        def _update_other_linux(ret, data):
-            if data.get('other_linux') == 'Unknown' and 'other_linux' in ret:
-                # Don't update 'other_linux' to 'Unknown' if identified already
-                data.pop('other_linux')
-            ret.update(data)
-            return ret
-
-        self._release = 'RHEL'
-        self._reasons = {}
-        _dmesg = dmesg.linux_version if dmesg else dmesg
-        if not list(filter(None, [uname, _dmesg, rpms])):
-            # When uname, dmesg, and rpms are all unavailable
-            if osr or rhr:
-                # Use 'os-release' and 'redhat-release
-                ret = _from_os_release(osr) if osr else dict()
-                if ret.get('other_linux', 'RHEL') == 'RHEL':
-                    ret.update(_from_redhat_release(rhr)) if rhr else None
-                if ret.get('other_linux', 'RHEL') != 'RHEL':
-                    self._release = ret['other_linux']
-                    self._reasons = {'reason': 'NON-RHEL: os-release/redhat-release'}
-            else:
-                # Nothing means NON-RHEL
-                self._release = 'Unknown'
-                self._reasons = {'reason': 'Nothing available to check'}
-        else:
-            # Uname -> Dmesg -> RPMs
-            result = _from_uname(uname) if uname else dict()
-            if dmesg and dmesg.linux_version:
-                result.update(_update_other_linux(result, _from_dmesg(dmesg)))
-            if rpms:
-                result.update(_update_other_linux(
-                                result, _from_installed_rpms(rpms, uname)))
-            # 'other_linux' means NON-RHEL
-            if 'other_linux' in result and result['other_linux'] != 'RHEL':
-                self._release = result.pop('other_linux')
-                self._reasons = result
+
+        def __identify_non_rhel(ret):
+            if ret and 'other_linux' in ret:
+                self._release = ret.pop('other_linux')
+                self._reasons = ret
+                if self._release == 'Unknown':
+                    return None  # Continue to identify when unknown
+                return self._release != 'RHEL'
+
+        self._release = 'Unknown'
+        self._reasons = dict(reason='Nothing available to check')
+
+        # Check installed_rpms first
+        # Only when installed_rpms is unavailable, check below specs with order
+        # - uname, dmesg, os_release and redhat_release
+        ret = _from_installed_rpms(rpms, uname)
+
+        if __identify_non_rhel(ret) is None:
```

Comment:
Here it should continue checking when the `other_linux` is unknown, not only None, right?

Reply:
Yup, See L265, this internal function returns `None` only when other_linux is set as `Unknown`, I would add a comment here.

Reply:
Got it. It looks good to me.